src/custom/prelude/array.rs

```diff
@@ -0,0 +1,638 @@
+//! Codegen for prelude array operations.
+use anyhow::{anyhow, Ok, Result};
+use hugr::{
+    extension::{
+        prelude::{array_type, option_type, ArrayOp, ArrayOpDef},
+        simple_op::MakeRegisteredOp,
+    },
+    ops::DataflowOpTrait as _,
+    types::TypeEnum,
+    HugrView,
+};
+use inkwell::{
+    builder::{Builder, BuilderError},
+    types::BasicType,
+    values::{ArrayValue, BasicValue as _, BasicValueEnum, IntValue, PointerValue},
+    IntPredicate,
+};
+
+use crate::{
+    emit::{EmitFuncContext, RowPromise},
+    sum::LLVMSumType,
+    types::{HugrType, TypingSession},
+};
+
+use super::PreludeCodegen;
+
+/// Helper function to allocate an array on the stack and pass a pointer to it
+/// to a closure.
+///
+/// The pointer forwarded to the closure is a pointer to the first element of
+/// the array.
+fn with_array_alloca<'c, T, E: From<BuilderError>>(
+    builder: &Builder<'c>,
+    array: ArrayValue<'c>,
+    go: impl FnOnce(PointerValue<'c>) -> Result<T, E>,
+) -> Result<T, E> {
+    let array_ty = array.get_type();
+    let array_len: IntValue<'c> = {
+        let ctx = builder.get_insert_block().unwrap().get_context();
+        ctx.i32_type().const_int(array_ty.len() as u64, false)
+    };
+    let ptr = builder.build_array_alloca(array_ty.get_element_type(), array_len, "")?;
+    let array_ptr = builder
+        .build_bitcast(ptr, array_ty.ptr_type(Default::default()), "")?
+        .into_pointer_value();
+    builder.build_store(array_ptr, array)?;
+    go(ptr)
+}
+
+pub fn emit_array_op<'c, H: HugrView>(
+    pcg: &impl PreludeCodegen,
+    ctx: &mut EmitFuncContext<'c, H>,
+    op: ArrayOp,
+    inputs: Vec<BasicValueEnum<'c>>,
+    outputs: RowPromise<'c>,
+) -> Result<()> {
+    let builder = ctx.builder();
+    let ts = ctx.typing_session();
+    let sig = op.clone().to_extension_op().unwrap().signature();
+    let ArrayOp {
+        def,
+        ref elem_ty,
+        size,
+    } = op;
+    let llvm_array_ty = pcg
+        .array_type(&ts, ts.llvm_type(elem_ty)?, size)
+        .as_basic_type_enum()
+        .into_array_type();
+    match def {
+        ArrayOpDef::new_array => {
+            let mut array_v = llvm_array_ty.get_undef();
+            for (i, v) in inputs.into_iter().enumerate() {
+                array_v = builder
+                    .build_insert_value(array_v, v, i as u32, "")?
+                    .into_array_value();
+            }
+            outputs.finish(builder, [array_v.as_basic_value_enum()])
+        }
+        ArrayOpDef::get => {
+            let [array_v, index_v] = inputs
+                .try_into()
+                .map_err(|_| anyhow!("ArrayOpDef::get expects two arguments"))?;
+            let array_v = array_v.into_array_value();
+            let index_v = index_v.into_int_value();
+            let elem = with_array_alloca(builder, array_v, |ptr| {
+                let elem_addr = unsafe { builder.build_gep(ptr, &[index_v], "")? };
```

Comment:
Should we aim for using `build_in_bounds_gep`?

Reply:
Fair enough if it doesn't fit our use case, but I'd expect undefined behaviour if we get the indices wrong!